pytorch_lightning/core/lightning.py

```diff
@@ -263,9 +264,13 @@ def _apply_batch_transfer_handler(
         self, batch: Any, device: Optional[torch.device] = None, dataloader_idx: int = 0
     ) -> Any:
         device = device or self.device
-        batch = self.on_before_batch_transfer(batch, dataloader_idx)
-        batch = self.transfer_batch_to_device(batch, device, dataloader_idx)
-        batch = self.on_after_batch_transfer(batch, dataloader_idx)
+        datahook_source = (
+            _DataHookSource(self) if self.trainer is None else self.trainer._data_connector._datahook_source
```

Comment:
can we do this in a way without adding more dependencies to referring to `self.trainer` inside of the LightningModule? 

Why is `_apply_batch_transfer_handler` defined in the LightningModule at all at this point? Should this logic be in the Trainer, so it can select which datasource to call these hooks on?

Reply:
we use `_apply_batch_transfer_handler` at other places like model_summary, to_torchscript, to_onnx. so here I added this check to make sure it doesn't rely on the trainer completely. If a trainer is not attached, it will use its own logic to move the data around.